pkg/windows/calico.go

```diff
@@ -60,25 +62,47 @@ func (c *Calico) Setup(ctx context.Context, dataDir string, nodeConfig *daemonco
 		return nil, err
 	}
 
-	secrets, err := client.CoreV1().Secrets(CalicoSystemNamespace).List(ctx, metav1.ListOptions{})
+	sa, err := serviceaccount.GetOrCreateServiceAccount(client.CoreV1(), CalicoSystemNamespace, calicoNode)
 	if err != nil {
 		return nil, err
 	}
 
-	for _, secret := range secrets.Items {
-		if !isCalicoNodeToken(&secret) {
-			continue
-		}
-		value, ok := secret.Data["token"]
-		if !ok {
-			continue
-		}
-		calicoKubeConfig.Token = string(value)
+	t := true
+	sa.AutomountServiceAccountToken = &t
+
+	saClient := client.CoreV1().ServiceAccounts(CalicoSystemNamespace)
+
+	req := authv1.TokenRequest{
+		TypeMeta:   sa.TypeMeta,
+		ObjectMeta: sa.ObjectMeta,
+		Spec: authv1.TokenRequestSpec{
+			BoundObjectRef: &authv1.BoundObjectReference{
+				Kind:       sa.Kind,
+				APIVersion: sa.APIVersion,
+				Name:       sa.Name,
+			},
+		},
 	}
-	if calicoKubeConfig.Token == "" {
-		return nil, errors.New("could not retrieve calico node token from the cluster")
+
+	token, err := saClient.CreateToken(ctx, sa.Name, &req, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	tokenStatus := token.Status
+	if tokenStatus.Token == "" {
+		return nil, fmt.Errorf("[Setup] calico-node service account (%s) returned an empty token (%s) from the cluster", sa.Name, token.Name)
+	}
+
+	if tokenStatus.ExpirationTimestamp.Time.Before(time.Now()) {
+		logrus.Warnf("[Setup] service account token (%s) is expired, attempting to generate a new one", token.Name)
+		token, err = saClient.CreateToken(ctx, sa.Name, &req, metav1.CreateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("[Setup] unable to create new token (%s) for service account (%s): %v", token.Name, sa.Name, err)
+		}
```

Comment:
Tokens aren't stored anywhere server-side. You get a new token for the SA every time you call CreateToken, so checking to see if it's expired right after retrieving it should be unnecessary.
```suggestion
```

 The docs do mention checking the expiration to see if the expiration is less than was requested in the `expirationSeconds` on the TokenRequest, which it looks like we're not setting? It would probably be good to manually validate that the returned tokens aren't excessively short in duration, such that they might expire while Calico is still trying to use them.